phantom_snap/render-ipc.js

```diff
@@ -1,233 +1,254 @@
+ *   "paintTime": Long of amount of time required to paint the page (render the image),
+ *   "base64": String of base 64 encoded image,
+ *   "format": String
+ * }\n
+
+ * In the event of an error, the output will be JSON:
+ * {
+ *   "error": String
+ * }\n
+ *
+ * To exit the process send:
+ * exit\n
+ */
+
+var system = require("system");
+var webpage = require("webpage");
+
+// Redirect console logging to stderr to protect the protocol on stdin/stdout
+console.log = function(msg) {
+    system.stderr.writeLine(msg);
+};
+
+// Redirect console logging to stderr to protect the protocol on stdin/stdout
+console.error = function(msg) {
+    system.stderr.writeLine(msg);
+};
+
+// From https://stackoverflow.com/a/41713002, https://jsfiddle.net/47zwb41o/
+escape = function( s ){
+    var chr, hex, i = 0, l = s.length, out = '';
+    for( ; i < l; i ++ ){
+        chr = s.charAt( i );
+        if( chr.search( /[A-Za-z0-9\@\*\_\+\-\.\/]/ ) > -1 ){
+            out += chr; continue; }
+        hex = s.charCodeAt( i ).toString( 16 );
+        out += '%' + ( hex.length % 2 != 0 ? '0' : '' ) + hex;
+    }
+    return out;
+};
+
+b64ToUtf8 = function(s){
+	s = s.replace(/\s/g, '');
+	return decodeURIComponent(escape(atob(s)));
+};
+
+renderHtml = function (request) {
+
+    var page = webpage.create();
+
+    page.viewportSize = {
+        width: 1280,
+        height: 1024
+    };
+
+    if (request.hasOwnProperty('width')) {
+        page.viewportSize.width = request.width;
+    }
+
+    if (request.hasOwnProperty('height')) {
+        page.viewportSize.height = request.height;
+    }
+
+    if (request.hasOwnProperty('userAgent')) {
+        page.settings.userAgent = request.userAgent;
+    }
+
+    if (request.hasOwnProperty('headers')) {
+        page.customHeaders = request.headers;
+    }
+
+    phantom.clearCookies();
+
+    if (request.hasOwnProperty('cookies')) {
+        for (var i = 0; i < request.cookies.length; i++) {
+            phantom.addCookie(request.cookies[i]);
+        }
+    }
+
+    var format = 'PNG';
+
+    if (request.hasOwnProperty('format')) {
+        if(request.format === 'PNG' || request.format === 'JPEG') {
+            format = request.format;
+        }
+        else {
+            var error = {
+                error: 'Request JSON specifies unsupported output format.'
+            };
+            system.stdout.writeLine(JSON.stringify(error));
+            readInput();
+        }
+    }
+
+    var timeout = 1000 * 180;
+    var resourceWait  = 300,
+    maxRenderWait = 10000;
+
+    var count = 0,
+    forcedRenderTimeout,
+    renderTimeout;
+
+    if(request.hasOwnProperty('timeout')) {
+        timeout = request.timeout;
+    }
+
+    page.settings.resourceTimeout  = timeout;
+
+    var rendered = false;
+    var time = Date.now();
+
+    page.onConfirm = page.onPrompt = function noOp() {};
+
+    var render = function(status) {
+
+        if(!rendered) {
```

Comment:
A logical weakness I'm worried about is with multiple scheduled render functions. When we do render, we should clear all of them to prevent a mis-timed execution across multiple requests, because a "stopped" render timeout could still potentially fire during the processing of a subsequent request.
```suggestion
        if(!rendered) {
          clearTimeout(renderTimeout);
          clearTimeout(stoppedRenderTimeout); // <--- New, needs to be assigned at line 215
```

Reply:
Why would there be multiple scheduled renders in a lambda function? That isn't how lambda functions work.

Reply:
Request A comes in. Rendering is scheduled for the max timeout (Rmax-a) and also for resource wait timeout (Rwait-a). Soon after, Rwait-a fires and the page is rendered and the response returned. PhantomJS (the process) is still running this script waiting for more activity on stdin. Next, request B comes in and sets up Rmax-b and Rwait-b. The page for request B has started loading and BAM!!! Rmax-a fires and preemptively renders the page, leaving Rmax-b and potentially Rwait-b to strike request C.

Keep in mind that this script is used outside Lambda too, with the local PhantomJS executable.

Reply:
Maybe the way javascript works with closures and scope this isn't an issue, and things like the `rendered` flag are unique on the stack for each request. Honestly, I don't know for sure.

Reply:
gotcha. done.